### PR TITLE
Add start_time column to sys.servers 

### LINKED ISSUE #12090: Add a service start_time column, so you know when each of your services was started.
### Description

It would be nice to track server start time for each of the services that come up. 

### Motivation

Server start time is an important variable when it comes to deployments and patching code. 

Please provide the following for the desired feature or change:

Add a start_time column in sys.servers or zk.  I haven't looked at the code, but it can be a default value.  I can look into keeping this backwards compatible after folks agree this is a good idea.

### PR DESCRIPTION
**Pull request by a2l007**
Fixes #12090

### Description

Adds a new column `start_time` to sys.servers that captures the time at which the server was added to the cluster.

<hr>

##### Key changed/added classes in this PR
 * `DiscoveryDruidNode`
 * `SystemSchema`

<hr>

This PR has:

- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: server/src/main/java/org/apache/druid/discovery/DiscoveryDruidNode.java [hunk 1]
```diff
@@ -164,6 +179,12 @@ public DruidNode getDruidNode()
     return druidNode;
   }
 
+  @JsonProperty
+  public DateTime getStartTime()
+  {
+    return startTime;
+  }
```
Comment (LakshSingla) [inline server/src/main/java/org/apache/druid/discovery/DiscoveryDruidNode.java:186]: Why is startTime omitted while verifying the equality of this class?

**Review by LakshSingla — COMMENTED** [review]
Thanks for the contribution! The Java code changes LGTM barring a minor question.

**Comment by vogievetsky** [comment]
Oh damn! so cool! I can not wait to add an "uptime" column to the web console!

File: web-console/src/views/services-view/services-view.tsx [hunk 2]
```diff
@@ -67,14 +67,25 @@ const allColumns: string[] = [
   'Current size',
   'Max size',
   'Usage',
+  'Start Time',
```
Comment (vogievetsky) [inline web-console/src/views/services-view/services-view.tsx:70]: Please capitalize as `Start time`

File: web-console/src/views/services-view/services-view.tsx [hunk 3]
```diff
@@ -510,6 +524,14 @@ ORDER BY
               }
             },
           },
+          {
+            Header: 'Start Time',
```
Comment (vogievetsky) [inline web-console/src/views/services-view/services-view.tsx:528]: Ditto re: capitalization

File: web-console/src/views/services-view/services-view.tsx [hunk 4]
```diff
@@ -211,6 +224,7 @@ ORDER BY
         curr_size: s.currSize,
         max_size: s.maxSize,
         tls_port: -1,
+        start_time: s.start_time,
```
Comment (vogievetsky) [inline web-console/src/views/services-view/services-view.tsx:227]: did you add `start_time` to the `/druid/coordinator/v1/servers?simple` response also? If so you should update https://github.com/apache/druid/blob/master/docs/operations/api-reference.md#L496 also, if not then update the line of code above.

File: server/src/main/java/org/apache/druid/discovery/DiscoveryDruidNode.java (hunk 1 above)
Comment (a2l007) [reply server/src/main/java/org/apache/druid/discovery/DiscoveryDruidNode.java:186]: Since a DiscoveryDruidNode object is primarily identified by its DruidNode, role and the service map, I wanted to preserve the equality condition. Also, start_time might not be a concrete enough to decide the equality between two  DiscoveryDruidNode objects, if the other field values are the same. What do you think?

File: web-console/src/views/services-view/services-view.tsx (hunk 4 above)
Comment (a2l007) [reply web-console/src/views/services-view/services-view.tsx]: Good catch, removed it.

File: server/src/main/java/org/apache/druid/discovery/DiscoveryDruidNode.java (hunk 1 above)
Comment (LakshSingla) [reply server/src/main/java/org/apache/druid/discovery/DiscoveryDruidNode.java:186]: Thanks for the explanation! I think it's better to keep it the current way.

File: web-console/src/views/services-view/services-view.tsx [hunk 5]
```diff
@@ -67,14 +67,25 @@ const allColumns: string[] = [
...
   'Max size',
   'Usage',
+  'Start time',
   'Detail',
   ACTION_COLUMN_LABEL,
 ];
 
 const tableColumns: Record<CapabilitiesMode, string[]> = {
   'full': allColumns,
   'no-sql': allColumns,
```
Comment (vogievetsky) [inline web-console/src/views/services-view/services-view.tsx:77]: ok, so since you are not updating `/druid/coordinator/v1/servers?simple` you should make sure that `Start time` is no in the `no-sql` column list. `no-sql` is the mode used when the user had no SQL access so it falls back to the old endpoint. You should set `no-sql` to:

```
[
  'Service',
  'Type',
  'Tier',
  'Host',
  'Port',
  'Current size',
  'Max size',
  'Usage',
  'Detail',
  ACTION_COLUMN_LABEL,
];
```

Also at this point you can inline the `allColumns` constant as it's reasons for existence was to set `full` and `no-sql` to the same thing.

**Review by vogievetsky — APPROVED** [review]
👍 for everything but the Java (I did not review the Java code - only TS + general idea). Thank you for promptly responding to feedback!

**Comment by a2l007** [comment]
@vogievetsky Thanks for the review. Do you know what is going on with the Travis job failure: web console end-to-end test ?
It builds fine locally but Travis seems to keep failing that job.

**Comment by vogievetsky** [comment]
not sure, will have a look in a bit.

**Review by abhishekagarwal87 — APPROVED** [review]

**Comment by a2l007** [comment]
I found an issue when coordinator starts up with `druid.coordinator.asOverlord.enabled` as true. In this case, the coordinator and overlord services are announced twice and in each case with a different instance of `DiscoveryDruidNode`. [This breaks the announcer flow](https://github.com/apache/druid/blob/master/server/src/main/java/org/apache/druid/curator/announcement/Announcer.java#L317) as each of the DiscoveryDruidNode objects have a slightly different `startTime` which causes a mismatch between the node bytes announced at the path. 
Converting this PR to a draft until I find a better way to fix this.

**Comment by vogievetsky** [comment]
Was that what was causing the e2e failures?

**Comment by a2l007** [comment]
> Was that what was causing the e2e failures?

@vogievetsky I believe so, since we run coordinator with asOverlord enabled in our builds.

**Comment by a2l007** [comment]
@vogievetsky Sorry for the delay, I'll find some time this week to fix up this PR

**Comment by a2l007** [comment]
@vogievetsky @abhishekagarwal87 @LakshSingla Sorry for the delay in fixing up the conflicts for this PR, but it'd be great if you could take a quick look at this again.

File: server/src/main/java/org/apache/druid/curator/announcement/Announcer.java [hunk 6]
```diff
@@ -314,7 +313,7 @@ public void childEvent(CuratorFramework client, PathChildrenCacheEvent event) th
         if (oldBytes == null) {
           created = true;
         } else if (!Arrays.equals(oldBytes, bytes)) {
-          throw new IAE("Cannot reannounce different values under the same path");
+          log.error("Ignoring attempt to announce different values under same path");
```
Comment (abhishekagarwal87) [inline server/src/main/java/org/apache/druid/curator/announcement/Announcer.java:316]: what is the rationale behind this change?

**Comment by abhishekagarwal87** [comment]
@a2l007 - the PR is ready to go. I just had one question on a change that you made in this PR.

File: server/src/main/java/org/apache/druid/curator/announcement/Announcer.java (hunk 6 above)
Comment (a2l007) [reply server/src/main/java/org/apache/druid/curator/announcement/Announcer.java:316]: When the coordinator is run in overlord mode, Announcer.announce() is called twice since the Announcer module is part of the coordinator and overlord lifecycle modules. The second call is a no-op since the existing DiscoveryDruidNode bytes announced at the path is the same as the node bytes in the second call.
With this patch, the start time is now part of DiscoveryDruidNode and so in some cases, there could be a millisecond delay between the two announce calls. This causes the node objects to be different and the second announce call fails due to the validation check.
I couldn't find another scenario where it would be useful to fail the process in this condition and so I'm logging it here instead. Let me know if you have any thoughts on this approach.

File: docs/querying/sql-metadata-tables.md [hunk 7]
```diff
@@ -208,7 +208,7 @@ Servers table lists all discovered servers in the cluster.
 |current_size|BIGINT|Current size of segments in bytes on this server. Only valid for HISTORICAL type, for other types it's 0|
 |max_size|BIGINT|Max size in bytes this server recommends to assign to segments see [druid.server.maxSize](../configuration/index.md#historical-general-configuration). Only valid for HISTORICAL type, for other types it's 0|
 |is_leader|BIGINT|1 if the server is currently the 'leader' (for services which have the concept of leadership), otherwise 0 if the server is not the leader, or the default long value (0 or null depending on `druid.generic.useDefaultValueForNull`) if the server type does not have the concept of leadership|
-
+|start_time|STRING|Timestamp in ISO8601 format corresponding to when the server was announced in the cluster|
```
Comment (techdocsmith) [inline docs/querying/sql-metadata-tables.md:211]:
```suggestion
|start_time|STRING|Timestamp in ISO8601 format when the server was announced in the cluster|
```

File: server/src/main/java/org/apache/druid/curator/announcement/Announcer.java (hunk 6 above)
Comment (abhishekagarwal87) [reply server/src/main/java/org/apache/druid/curator/announcement/Announcer.java:316]: it's called 4 times. From my local logs of a run
```
2023-03-22T13:24:51,527 INFO [main] org.apache.druid.curator.discovery.CuratorDruidNodeAnnouncer - Announced self [{"druidNode":{"service":"druid/coordinator","host":"localhost","bindOnHost":false,"plaintextPort":8081,"port":-1,"tlsPort":-1,"enablePlaintextPort":true,"enableTlsPort":false},"nodeType":"coordinator","services":{}}].
2023-03-22T13:24:51,530 INFO [main] org.apache.druid.curator.discovery.CuratorDruidNodeAnnouncer - Announced self [{"druidNode":{"service":"druid/coordinator","host":"localhost","bindOnHost":false,"plaintextPort":8081,"port":-1,"tlsPort":-1,"enablePlaintextPort":true,"enableTlsPort":false},"nodeType":"overlord","services":{}}].
2023-03-22T13:24:51,530 INFO [main] org.apache.druid.curator.discovery.CuratorDruidNodeAnnouncer - Announced self [{"druidNode":{"service":"druid/coordinator","host":"localhost","bindOnHost":false,"plaintextPort":8081,"port":-1,"tlsPort":-1,"enablePlaintextPort":true,"enableTlsPort":false},"nodeType":"coordinator","services":{}}].
2023-03-22T13:24:51,530 INFO [main] org.apache.druid.curator.discovery.CuratorDruidNodeAnnouncer - Announced self [{"druidNode":{"service":"druid/coordinator","host":"localhost","bindOnHost":false,"plaintextPort":8081,"port":-1,"tlsPort":-1,"enablePlaintextPort":true,"enableTlsPort":false},"nodeType":"overlord","services":{}}].
```

I debugged this a bit and you are right about it being called from two modules. I think that we can also skip the duplicate announcement when Overlord is not in standalone mode. That should fix the problem you ran into. And we keep this assert in place. what do you think?

Comment (abhishekagarwal87) [reply server/src/main/java/org/apache/druid/curator/announcement/Announcer.java:316]: I tried this change and it's working fine. 

```
diff --git a/services/src/main/java/org/apache/druid/cli/CliOverlord.java b/services/src/main/java/org/apache/druid/cli/CliOverlord.java
index e4383c673c..79b90e63b5 100644
--- a/services/src/main/java/org/apache/druid/cli/CliOverlord.java
+++ b/services/src/main/java/org/apache/druid/cli/CliOverlord.java
@@ -267,13 +267,13 @@ public class CliOverlord extends ServerRunnable
 
             if (standalone) {
               LifecycleModule.register(binder, Server.class);
-            }
 
-            bindAnnouncer(
-                binder,
-                IndexingService.class,
-                DiscoverySideEffectsProvider.create()
-            );
+              bindAnnouncer(
+                      binder,
+                      IndexingService.class,
+                      DiscoverySideEffectsProvider.create()
+              );
+            }
 
             Jerseys.addResource(binder, SelfDiscoveryResource.class);
             LifecycleModule.registerKey(binder, Key.get(SelfDiscoveryResource.class));

```

Comment (a2l007) [reply server/src/main/java/org/apache/druid/curator/announcement/Announcer.java]: Thanks, that seems like the better solution. I've tested out the changes and it works as expected.

**Comment by abhishekagarwal87** [comment]
thank you @a2l007. I merged this PR. This missed the cut for the 26 milestone. if you want this in 26 release, please create a backport PR against 26.0.0 branch.
